### PR TITLE
Add h5py wheel

### PR DESCRIPTION
**Pull request by chambm**
It depends on libhdf5, so it might not be very portable.

### CODE REVIEW

**Comment by chambm** [comment]
@natefoo What next?

**Comment by martenson** [comment]
@galaxybot test this

File: wheels/build/wheels.yml [hunk 1]
```diff
@@ -15,6 +15,14 @@ packages:
         insert_setuptools: true
     numpy:
         version: 1.9.2
+    h5py:
+        version: 2.5.0
+        prebuild: "git clone -b linux-wheels https://github.com/natefoo/pip && cd pip && for py in /python/cp*-`uname -m`; do ${py}/bin/python setup.py install && ${py}/bin/pip install --index-url https://wheels.galaxyproject.org/simple/ Cython numpy || exit 1; done"
+        src:
+            - https://pypi.python.org/packages/source/h/h5py/h5py-2.5.0.tar.gz
```
Comment (natefoo) [inline wheels/build/wheels.yml]: The `src` section should be unnecessary, if left out, Starforge will download the sdist from PyPI automatically.

**Comment by natefoo** [comment]
@chambm in addition to removing `src`, could you add:

``` yaml
         brew:
            - homebrew/science/hdf5
```

Then we can rerun the Jenkins build and see how it goes.

The initial attempt at building on OS X failed (despite being reported as success), although the Linux builds did succeed. The error occurred while installing pip in the prebuild action:

``` pytb
Traceback (most recent call last):
  File "/python/cp26m-x86_64/bin/pip", line 9, in <module>
    load_entry_point('pip==8.0.0.dev0', 'console_scripts', 'pip')()
  File "build/bdist.macosx-10.6-intel/egg/pkg_resources/__init__.py", line 558, in load_entry_point
  File "build/bdist.macosx-10.6-intel/egg/pkg_resources/__init__.py", line 2682, in load_entry_point
  File "build/bdist.macosx-10.6-intel/egg/pkg_resources/__init__.py", line 2355, in load
  File "build/bdist.macosx-10.6-intel/egg/pkg_resources/__init__.py", line 2361, in resolve
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/__init__.py", line 13, in <module>
    from pip.utils import get_installed_distributions, get_prog
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/utils/__init__.py", line 18, in <module>
    from pip.compat import console_to_str, expanduser, stdlib_pkgs
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/compat/__init__.py", line 19, in <module>
    from pip._vendor import ipaddress
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/_vendor/ipaddress.py", line 1671, in <module>
    class _IPv4Constants(object):
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/_vendor/ipaddress.py", line 1673, in _IPv4Constants
    _linklocal_network = IPv4Network('169.254.0.0/16')
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/_vendor/ipaddress.py", line 1639, in __init__
    self.network_address = IPv4Address(self._ip_int_from_string(addr[0]))
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/_vendor/ipaddress.py", line 1387, in __init__
    self._check_int_address(address)
  File "/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/site-packages/pip-8.0.0.dev0-py2.6.egg/pip/_vendor/ipaddress.py", line 536, in _check_int_address
    self._version))
pip._vendor.ipaddress.AddressValueError: 2851995648 (>= 2**32) is not permitted as an IPv4 address
Traceback (most recent call last):
  File "/python/wheelenv/bin/starforge", line 9, in <module>
    load_entry_point('starforge==0.0.1.dev0', 'console_scripts', 'starforge')()
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/python/wheelenv/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/python/wheelenv/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/python/wheelenv/src/starforge/starforge/commands/cmd_bdist_wheel.py", line 65, in cli
    forge.bdist_wheel(output=output, uid=uid, gid=gid)
  File "/python/wheelenv/src/starforge/starforge/forge/wheels.py", line 171, in bdist_wheel
    root = self._prep_build(build, output, uid, gid)
  File "/python/wheelenv/src/starforge/starforge/forge/wheels.py", line 126, in _prep_build
    subprocess.check_call(prebuild, shell=True)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'git clone -b linux-wheels https://github.com/natefoo/pip && cd pip && for py in /python/cp*-`uname -m`; do ${py}/bin/python setup.py install && ${py}/bin/pip install --index-url https://wheels.galaxyproject.org/simple/ Cython numpy || exit 1; done' returned non-zero exit status 1
```

This fairly insane error occurred **because** `2**32 == 0` **in Python 2.6 in the OS X image** (and also amusingly, `2**31 == -2147483648`).

For the sake of history, the args used for building Python 2.6 were taken from the `lib/python2.6/config/Makefile` of a binary-package-installed 2.6 on an old system, back when python.org used to provide OS X installers for Intel-fat Python for targeting 10.6. Since these no longer exist, we compile our own Intel-fat Framework Python in the same fashion as the old installer versions. The offending argument was the `-O3` CFLAG, which was probably fine in older gcc versions on older OS X versions, but must cause some interesting differences in the underlying C types of Python types in llvm on Yosemite. Removing `-O3` results in a much less broken Python.

I fixed the OS X image in f2054a870bae0455269588e57300a97ba22530f1 and rebuilt it on mjolnir0 (the Starforge build slave). I tested using starforge to build the `h5py` wheels on this image (with the above changes to `wheels.py`) and it was successful.

**Comment by natefoo** [comment]
@galaxybot test this

**Comment by natefoo** [comment]
Grr... missing authorized_keys in the image. =/

**Comment by martenson** [comment]
at least it reports fail now :)

**Comment by natefoo** [comment]
@galaxybot test this

**Comment by natefoo** [comment]
Latest starforge should probably be installed on mjolnir0 automatically, right now it's done by hand.

**Comment by natefoo** [comment]
@galaxybot test this

**Comment by natefoo** [comment]
https://depot.galaxyproject.org/starforge/wheels/build-19/

**Comment by martenson** [comment]
yay! I think we can mak galaxybot comment with that link

**Comment by natefoo** [comment]
@martenson That'd be great.

**Comment by bgruening** [comment]
So I guess we can merge this?

**Comment by natefoo** [comment]
Yes and no.

@chambm would like to see libhd5.so bundled into the wheel so that we do not have to expect people to have libhdf5 installed on their systems to use the h5py wheel. I guess we should decide whether or not to do that. It's sort of the situation I was trying to avoid e.g. with the psycopg2 wheels, but in the case of psycopg2, I prefer to rely on the system version for client/server compatibility (although this never ended up being a problem with the psycopg2 egg) and security reasons. This may be less of an issue with libhd5 however.

distutils-sig discussed this a few months ago as well, but I don't recall where the conversation ended up.

**Comment by jmchilton** [comment]
With libhdf5 in Galaxy, I'd be more concerned about reproduciblity and ease of use than security. I'd vote to include it.

**Comment by chambm** [comment]
Without libhdf5 included, there's not much point including h5py. If a user can install libhdf5 (i.e. they have sudo or ask their admin to do it) then they can also install python-h5py. Having both included would allow people to get it without messing with external package managers and/or pesky administrators. ;)

**Comment by bgruening** [comment]
:+1: for including

**Comment by natefoo** [comment]
I'll work on getting starforge to support this.

**Comment by bgruening** [comment]
Any news on this @natefoo?
Sorry for all the trouble :(

**Comment by natefoo** [comment]
Yes! This can now easily be done with manylinux1 wheels.

**Comment by jmchilton** [comment]
Closing in favor #145 I guess.

**Pull request closed without merging**